src/main/ruby/truffleruby/core/integer.rb

```diff
@@ -332,6 +332,19 @@ def zero?
     self == 0
   end
 
+  def self.try_convert(obj)
```

Comment:
I think we should reuse one of these methods under `Truffle::Type`: https://github.com/oracle/truffleruby/blob/fac6bcf8d04e70dd4f6ccb09f4ddd71832824e9a/src/main/ruby/truffleruby/core/type.rb#L237-L252 or https://github.com/oracle/truffleruby/blob/fac6bcf8d04e70dd4f6ccb09f4ddd71832824e9a/src/main/ruby/truffleruby/core/type.rb#L354-L362 to avoid code duplication and make it easier to keep it consistent and maintain.

Reply:
Those methods seem to do what we want, but they raise TypeError with different messages when conversion fails, so Integer.try_convert should rescue TypeError and re-raise with the required message. 

The problem is that Integer.try_convert includes the type of the result of calling :to_int in the error message, which can't be extracted from those helper methods when they raise exceptions. (well, it can, but in a very ugly way, involving string manipulation of the exception's error message) 

All in all, I think duplication is okay here, the APIs are subtly mis-matching.

Reply:
```
irb(main):001:1>       class DifferentMockObject
irb(main):002:2>         def to_int
irb(main):003:2>           Object.new
irb(main):004:1>         end
irb(main):005:0>       end
=> :to_int
irb(main):006:0> obj=DifferentMockObject.new
irb(main):007:0> Truffle::Type.rb_check_convert_type obj, Integer, :to_int
<internal:core> core/type.rb:187:in `conversion_mismatch': can't convert DifferentMockObject to Integer (DifferentMockObject#to_int gives Object) (TypeError)                                            
```
`Truffle::Type.rb_check_convert_type` seems to give the right message, except the `to`->`into`, that's unfortunate inconsistency in CRuby.

So yeah either we duplicate or we accept both in ruby/spec (by using a Regexp).

Reply:
I think there is one way, we could adapt `Truffle::Type.conversion_mismatch`.
If the `cls` is Integer or Float, then we use `into`, and otherwise we use `to`.
That seems the overall distinction for to/into in CRuby behavior from a quick look:
```
irb(main):001:0> sprintf("%b", Object.new)
(irb):1:in `sprintf': can't convert Object into Integer (TypeError)
irb(main):008:0> obj=Object.new
=> #<Object:0x00007f0d201a0980>
irb(main):009:0> def obj.to_ary; "foo"; end
=> :to_ary
irb(main):010:0> Array.try_convert(obj)
(irb):10:in `try_convert': can't convert Object to Array (Object#to_ary gives String) (TypeError)
```

Could you try that?

Reply:
If that doesn't work, let's add `Truffle::Type.conversion_mismatch_into` which is a copy of `conversion_mismatch` but using `into` and use that in Integer.try_convert.

Reply:
Looking more into CRuby it's not just Integer/Float, sometimes `into` is also used for String, Rational, etc.
Let's just use add `Truffle::Type.conversion_mismatch_into` then.

Reply:
Added conversion_mismatch_into right below conversion_mismatch in the same file, and removed all the File.dirname operations into their own branch (but they left a bunch of irrelevant effects, like whitepsace-only commits and such, sorry!).

I thought about adding an optional argument to conversion_mismatch that controls how it prints the error message, but it looks very ugly, so I just accepted the duplication.

Reply:
Thanks, looks good. Local duplication like that is fine, it's easy to keep in sync. Duplication far apart from each other, that's much harder to maintain.